OrbitLinuxTracing/UprobesUnwindingVisitor.cpp

```diff
@@ -16,13 +16,16 @@ void UprobesUnwindingVisitor::visit(SamplePerfEvent* event) {
   const std::vector<unwindstack::FrameData>& full_callstack =
       unwinder_.Unwind(current_maps_.get(), event->GetRegisters(),
                        event->GetStackData(), event->GetStackSize());
-  if (!full_callstack.empty()) {
+  // Some samples can actually fall inside u(ret)probes code. Discard them,
```

Comment:
I'm wondering: Even if unwinding could be made successful, would these samples be useful? This is not game code, it is actually code that was artificially added by the profiler. If the u(ret)probe code was not there, the sample would be taken a "few lines further down", so probably either in the function that was called (in the case of a uprobe) or in the function that we would return into (in the case of a uretprobe), but we can't really know for sure where we would land exactly. So, I think discarding them is the right thing to do.

I'm also wondering if we should show this to the user in some way. First, we should probably count how many samples are discarded because of this issue. If it is a relatively large number, we should display to the user that many samples were discarded due to influence from instrumentation. Then they can decide how much to trust the sampling data. I'm not sure what a "relatively large number" here would be, but I could imagine if you instrument a function that is very short and called very often, then a relatively large percentage of samples could actually fall inside u(ret)probe code. @dpallotti Do you have any numbers on this?

Reply:
> would these samples be useful?

One way I could think they might be useful would be to highlight how a lot of time is spent in profiling-related code during profiling. But on the other hand they could instead be confusing as they are unexpected.

I added a counter for the discarded samples, which is very useful because I now have numbers (on extreme cases).
Instrumenting the first ~20 inclusive time and first ~20 exclusive time from the main binary of Infiltrator (this leads to around 250k calls to instrumented functions per second, so a lot of time is spent in u(ret)probes code), the discarded samples are around 7%.
With the function referred to in the bug (b/153026037), called around 125k times per second, the number of discarded samples is definitely very large, much more than expected, to almost half. But let's keep in mind that that is really a special case, with a very tiny and fundamental function whose time is severely inflated by uprobes. After all, instrumenting only that function causes the frame rate to plummet and the vast majority of the samples that are not discarded to fall inside that function anyway. So maybe discarding that many samples is not as bad as it seems.

Reply:
It's fine to discard that many samples (do we have a choice?), but we must make it obvious to a user of ORBIT that this is happening.